modules/ding_innskraning/ding_innskraning.module

```diff
@@ -0,0 +1,780 @@
+        '#title' => t('Log in via !service',
+                  array(
+                    '!service' => 'Island.is',
+                  )),
+        '#weight' => 100,
+      );
+    }
+  }
+}
+
+/**
+ * Implements hook_form_FORM_ID_alter() for profile2_form().
+ *
+ * Hide the pincode change widget for island.is authenticated users. This
+ * modules weight ensures we get called after
+ * ding_user_form_profile2_form_alter().
+ */
+function ding_innskraning_form_profile2_form_alter(&$form, &$form_state) {
+  if (ding_innskraning_authenticated()) {
+    if (isset($form['pass'])) {
+      $form['pass']['#access'] = FALSE;
+    }
+  }
+}
+
+/**
+ * Whether the current user was authenticated with island.is.
+ */
+function ding_innskraning_authenticated() {
+  try {
+    $creds = ding_user_get_creds();
+    if ($creds && isset($creds['external-auth'])) {
+      return TRUE;
+    }
+  }
+  catch (Exception $e) {
+    // Move along, nothing to see here...
+  }
+  return FALSE;
+}
+
+/**
+ * Create the login link for https://innskraning.island.is.
+ */
+function ding_innskraning_saml_login() {
+  libraries_load('xmlseclibs');
+
+  $provider_url = 'https://innskraning.island.is';
+  $url = $provider_url;
+  $id = variable_get('ding_innskraning_id', '');
+
+  // Start a session to store the token in, so we can check it on return. This
+  // guards against attackers redirecting victims to island.is with return urls
+  // that might cause trouble. Currently the damage might be limited (delting
+  // reservations?), but there might be more damaging things that can be done in
+  // the future.
+  drupal_session_start();
+  $token = _ding_innskraning_auth_id();
+  $_SESSION['ding_innskraning_token'] = $token;
+  $url .= "?id=$id&authid=$token";
+  watchdog('ding_innskraning', 'Redirect to SAML login url: @url', array('@url' => $url), WATCHDOG_INFO);
+
+  // Prepare the parameters to pass to island.is.
+  $parameters['id'] = $id;
+  $parameters['authid'] = $token;
+  // Create the request body.
+  $samlRequest = _ding_innskraning_login_request($token, $id, $provider_url);
+  $parameters['SAMLRequest'] = $samlRequest;
+  $parameters['RelayState'] = url('<front>', array('absolute' => TRUE));
+  // @TODO: why is this commented out.
+  // Sign the request.
+  /*$signature = _ding_innskraning_signature($samlRequest, $parameters['RelayState']);
+    $parameters['SigAlg'] = XMLSecurityKey::RSA_SHA1;
+    $parameters['Signature'] = $signature;*/
+
+  header('Pragma: no-cache');
+  header('Cache-Control: no-cache, must-revalidate');
+  drupal_goto($provider_url, array('query' => $parameters));
+}
+
+/**
+ * Consume authentications from https://innskraning.island.is.
+ */
+function ding_innskraning_saml_endpoint() {
+  // Extract token from request.
+  if (!array_key_exists('token', $_POST)) {
+    watchdog('ding_innskraning', "Post parameter 'token' not found in SAML response.", array(), WATCHDOG_WARNING);
+    drupal_goto('user/login');
+    return FALSE;
```

Comment:
return `FALSE` after `drupal_goto()`?

Reply:
Hah, I was blinded by the returns after throws. All removed.